File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// Problem 1: First Applied
export function firstThenApply<T, U>(
  items: T[],
  predicate: (item: T) => boolean,
  transform: (item: T) => U
): U | undefined {
  const firstItem = items.find(predicate)
  return firstItem !== undefined ? transform(firstItem) : undefined
}

// Problem 2: Powers Generator
export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let powerValue = 1n; ; powerValue *= base) {
    yield powerValue
  }
}

// Problem 3: Line Count
export async function meaningfulLineCount(filename: string): Promise<number> {
  let validLineCount = 0
  const fileHandle = await open(filename, "r")

  try {
    for await (const line of fileHandle.readLines()) {
      const strippedLine = line.trim()
      if (strippedLine && !strippedLine.startsWith("#")) {
        validLineCount++
      }
    }
  } catch (err) {
    throw new Error(`No such file: ${filename}`)
  } finally {
    await fileHandle.close()
  }
  return validLineCount
}

// Problem 4: Shapes Surface Area/Volume
interface Sphere {
  readonly kind: "Sphere"
  readonly radius: number
}

interface Box {
  readonly kind: "Box"
  readonly width: number
  readonly length: number
  readonly depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2
    case "Box":
      return (
        2 *
        (shape.length * shape.width +
          shape.width * shape.depth +
          shape.depth * shape.length)
      )
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * shape.radius ** 3
    case "Box":
      return shape.length * shape.width * shape.depth
  }
}

// Problem 5: Binary Search Tree
export interface BinarySearchTree<T> {
  size(): number
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  inorder(): Iterable<T>
}

export interface Comparable<T> {
  compareTo(other: T): number
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0
  }

  contains(value: T): boolean {
    return false
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty<T>(), new Empty<T>())
  }

  *inorder(): Iterable<T> {
    return
  }

  toString(): string {
    return "()"
  }
}

export class Node<T> implements BinarySearchTree<T> {
  constructor(
    private nodeValue: T,
    private leftChild: BinarySearchTree<T>,
    private rightChild: BinarySearchTree<T>
  ) {}

  size(): number {
    return 1 + this.leftChild.size() + this.rightChild.size()
  }

  contains(value: T): boolean {
    if (value < this.nodeValue) {
      return this.leftChild.contains(value)
    } else if (value > this.nodeValue) {
      return this.rightChild.contains(value)
    } else {
      return true
    }
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.nodeValue) {
      return new Node(
        this.nodeValue,
        this.leftChild.insert(value),
        this.rightChild
      )
    } else if (value > this.nodeValue) {
      return new Node(
        this.nodeValue,
        this.leftChild,
        this.rightChild.insert(value)
      )
    } else {
      return this
    }
  }

  *inorder(): Iterable<T> {
    yield* this.leftChild.inorder()
    yield this.nodeValue
    yield* this.rightChild.inorder()
  }

  toString(): string {
    const nodeStrings: string[] = []
    if (!(this.leftChild instanceof Empty)) {
      nodeStrings.push(this.leftChild.toString())
    }
    nodeStrings.push(String(this.nodeValue))
    if (!(this.rightChild instanceof Empty)) {
      nodeStrings.push(this.rightChild.toString())
    }
    return `(${nodeStrings.join("")})`
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

(* Problem 1 --First Applied *)
let first_then_apply array predicate consumer =
  match List.find_opt predicate array with
  | None -> None
  | Some matched_element -> consumer matched_element

(* Problem 2 -- Powers Generator *)
let powers_generator base = 
  let rec generate_from power () =
    Seq.Cons (power,generate_from (power * base))
  in
  generate_from 1

(* Problem 3 -- Line Count *)
let meaningful_line_count filename = 
  let meaningful_line line =
    let trimmed_line = String.trim line in
    String.length trimmed_line > 0 && not (String.starts_with ~prefix:"#" trimmed_line)
  in
  let input_file = open_in filename in
  let finally () = close_in input_file in
  let rec count_lines line_count =
    try
      let line = input_line input_file in
      if meaningful_line line then
        count_lines (line_count + 1)
      else 
        count_lines line_count
    with 
      End_of_file -> line_count
  in
  Fun.protect ~finally (fun () -> count_lines 0);;

(* Problem 4 -- Shapes Surface Area/Volume*)
type shape = 
  | Sphere of float
  | Box of float * float * float

let volume shape_item = 
  match shape_item with 
  | Sphere radius -> Float.pi *. (radius ** 3.) *. 4. /. 3.
  | Box (length, width, height) -> length *. width *. height

let surface_area shape_item = 
  match shape_item with
  | Sphere radius -> Float.pi *. (radius ** 2.) *. 4.
  | Box (length, width, height) -> 2. *. (length *. width +. height *. length +. height *. width)

(* Problem 5 -- Binary Search Tree *)
type 'a binary_search_tree = 
  | Empty 
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size tree = 
  match tree with 
  | Empty -> 0
  | Node (left_child, _, right_child) -> 1 + size left_child + size right_child

let rec insert value tree =
  match tree with
  | Empty -> Node(Empty, value, Empty)
  | Node (left_child, node_value, right_child) -> 
    if value < node_value then
      Node (insert value left_child, node_value, right_child)
    else if value > node_value then
      Node (left_child, node_value, insert value right_child)
    else 
      tree 

let rec contains value tree =
  match tree with
  | Empty -> false
  | Node (left_child, node_value, right_child) -> 
    if value < node_value then 
      contains value left_child
    else if value > node_value then
      contains value right_child 
    else
      true

let rec inorder tree = 
  match tree with 
  | Empty -> []
  | Node (left_child, node_value, right_child) -> inorder left_child @ [node_value] @ inorder right_child